### DISCUSSION TITLE
Non-diagonal Covariance

### DISCUSSION THREAD

**Discussion by DanielAmsel-fau** [voxelmorph/voxelmorph#256] [Q&A] [answered]
Hi!
In the paper "Unsupervised Learning of Probabilistic Diffeomorphic Registration for Images and Surfaces" in section 3.2 "Non-diagonal Covariance" you introduce an approximation for Sigma using a gaussian convolution resulting in a non-diagonal approximation of the Covariance. In the section "Derivation of Main Loss" you simplify the equation using some facts ending up with the final loss formula (12). One of those facts is tr((λD − A)Σz|m,f ) = tr(λDΣz|m,f ) (a). As far as I understood this equation (a) holds because Sigma is diagonal and A (adjacency matrix of Laplacian) is 0 along the diagonal, therefore resulting in 0 when multiplied. This would mean that (a) does not hold for the non-diagonal Covariance approximation, because A multiplied with Σz|m,f would in general not always result in 0. Considering this assumptions the loss formula (12) would be unsuitable for loss computition using the non-diagonal Covariance approximation.
Is this assumption right or did I understand something wrong?
I would be glad to get a reply to my question.

**Comment by adalca** [comment, marked as answer]
@DanielAmsel-fau 

It's been a while, but I believe you're correct in that this complicates the loss. However, computing the (original) trace is still achievable using convolutions (since we created Sigma in such a way as to enable that, as we describe in 3.2), which I think is how we did it in the implementation.

**Reply by DanielAmsel-fau** [reply]
Thank you for your answer :)
I will change the given implementation slightly to compute the original trace the way you mentioned it.

**Reply by DanielAmsel-fau** [reply]
While implementing the gaussian convolution another question came to my mind: How exactly do I choose the sigma for the gaussian kernel? If I simply rearrange the equation given for sigma in Section 3.2  I end up with fairly large values which lead to unwanted results.

**Reply by adalca** [reply]
Sorry I missed this -- which sigma are we talking here? the image noise std?

I will note a couple of general things -- some of these are hyperparameters that need tuning, which is annoying but fairly standard in registration and ML in general. We recently had a paper that addresses this, though: https://arxiv.org/abs/2101.01035

**Reply by DanielAmsel-fau** [reply]
With sigma I mean the variance σ_c  squared, which is used to compute the gaussian kernel for the gaussian convolution as described in Section 3.2 of the above mentioned paper. In the paper you provided a formula: 1/sqrt[2π(σ_c)^3/2] = (λ ∗ 6)^−1. Solving this equation for σ_c resolves in large values for the variance. For example solving this equation for the prior lambda you suggested in the README which is 25 would lead to a σ_c of roughly 234. Do I miss out on something regarding the computition of σ_c?

I will have a look at the paper aswell, thanks for that.

**Reply by adalca** [reply]
Okay, this is a typo (good spot!)

the formula should say: (omitting the inverses since they cancel out)
[sqrt(2pi) * sigma]^3 = lambda * 6
which i think leads to
sigma = 1/sqrt(2pi) [6 lambda] ^ (1/3)

so lambda of 20 (Fig 11) gives more reasonable sigmas.

I did this kind of quick but i think this is the right thing.

**Reply by DanielAmsel-fau** [reply]
Thank you for your help so far. I have another two questions for you.
1) While analyzing the Covariance-matrices of my model that's trained using probabilities, I noticed that the matrices are not diagonal. That obviously should not be the case. Is this a sign of my model not converging properly yet? I modified the pytorch-version of voxelmorph to support the probability-option aswell, therefore this could also be a bug in my code.

2) I would like to perform an uncertainty estimation for my model. Do you have any code for that? If not, an explanation of the formula given in Section 2.4 of the MICCAI paper would be great. In this case I will try to implement it myself.

**Reply by adalca** [reply]
@DanielAmsel-fau 

1. Well in our MICCAI implementation, we only predict the diagonal of the covariance, so I'm not quite sure I understand what you mean by having non-diagonal matrices?
 
2. I can't easily find this code right now, but it's fairly simple. To estimate uncertainty, in equation (8) of the MICCAI paper we approximate it via [entropy](https://en.wikipedia.org/wiki/Entropy_(information_theory)) via the approximate posterior, which amounts to using the covariance of the velocity field. If you want to covariance of the deformation field, you can sample velocity fields, integrate them, and then compute the covariance of those samples.

**Reply by DanielAmsel-fau** [reply]
About 1. I looked at the covariance matrices outputed by the U-Net. Those were not diagonal in my case. What exactly do you mean by predicting the diagonal of the covariance matrix? 
About 2. I was wondering about the index j used in the Section of the MICCAI paper mentioned above. What exaclty do z[j] and Σ[j, j] represent? Is j a coordinate vector (e.g. (x, y) in the 2D case)?

**Reply by adalca** [reply]
1. The image covariance would be N x N where N = HxWxDx3 for a 3D volume. A diagonal covariance would only have the N elements on the diagonal, and everything else 0. By construction, we only output those N diagonal elements from the UNet

2. `j` indexes into the matrix/vector. `z` is a vector of size N (or slightly smaller if we use smaller volumes for the velocity field, but for purposes of explanation think of it as size N, like the entire velocity field vectorized). The covariance is size N x N. So z[j] indexes the j'th element of z, \Sigma[j,j] is the jth element of the diagonal.

**Reply by DanielAmsel-fau** [reply]
Thank you again for your answers. Especially 1 cleared up a lot of things for me. I have another question that goes in a different direction: I want to train my probabilistic model using mini-batches. As far as I have reviewed and implemented the code of the KL-Loss it should be suited for training with batches. Unfortunately the results are far off the expected results that I achieve using batchsize 1. I tried playing around with the lambda hyperparameters but this did not quite work out for me. Do you have any suggestions regarding my problem?

**Reply by adalca** [reply]
We've done quite a bit of `vxm` training with higher batch sizes and it's worked well, although I'm not sure how much of that was with the probabilistic version. Can you tell me a bit more about batchsize=1 vs batchsize=2? How do the training curves look like?

**Reply by DanielAmsel-fau** [reply]
Training curves look pretty similar with the training loss for batchsize 2 being slightly higher. The main difference lies in the network output. The deformationfields for batchsize 2 indicate motion everywhere even though the backround is the same for the moving and the fixed image (use case is cardio MRI registration btw). For batchsize 1 the deformationfields are really smooth and there is only motion where it is expected to be.

**Reply by adalca** [reply]
hmm, sorry not quite sure what to make of this off the top of my head, and don't have time right now to look into it unfortunately. I'm wondering if we're accidentally summing instead of averaging by mis-using broadcasting, or something like that. Sorry, I'm not quite sure right now. I'll try to experiment on some simple data like MNIST at some point.

**Reply by DanielAmsel-fau** [reply]
I kinda forgot about this conversation, sorry for that. Training with batches works fine, the issue was the split into mu and sigma for batchsize output.

Thank you for all your answers, they helped a lot.

**Reply by adalca** [reply]
glad to hear!
